javatests/arcs/core/data/expression/TypeCheckerTest.kt

```diff
@@ -0,0 +1,479 @@
+  }
+
+  @Test
+  fun check_from_expressions() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val people = SeqType(ScopeType(emptyScope.builder().set("name", TextType).build()))
+    val scope =
+      emptyScope.builder().apply {
+        set("numbers", SeqType(IntType))
+        set("people", people)
+        set("company", SeqType(ScopeType(emptyScope.builder().set("people", people).build())))
+      }.build()
+
+    checkTypeIs(
+      PaxelParser.parse("from p in numbers select p"),
+      SeqType(IntType),
+      scope
+    )
+
+    checkTypeIs(
+      PaxelParser.parse("from p in people select p.name"),
+      SeqType(TextType),
+      scope
+    )
+
+    checkTypeIs(
+      PaxelParser.parse("from c in company from p in c.people select p.name"),
+      SeqType(TextType),
+      scope
+    )
+  }
+
+  @Test
+  fun check_where_expression() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val scope = emptyScope.builder().apply {
+      set("numbers", SeqType(IntType))
+      set("x", TextType)
+    }.build()
+
+    checkTypeIs(
+      PaxelParser.parse("from p in numbers where p < 10 select p"),
+      SeqType(IntType),
+      scope
+    )
+
+    checkTypeIs(
+      PaxelParser.parse("from p in numbers where 2 + 3 select p"),
+      SeqType(IntType),
+      scope,
+      errors = listOf("from p in numbers\nwhere 2.0 + 3.0 must evaluate to a boolean type" +
+                        " but was PrimitiveType(Double).")
+    )
+  }
+
+  @Test
+  fun check_select_expression() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val scope = emptyScope.builder().apply {
+      set("numbers", SeqType(IntType))
+    }.build()
+
+    checkTypeIs(
+      PaxelParser.parse("from p in numbers select new Foo { x: p, y: p + 1i, z: p == 2 }"),
+      SeqType(
+        ScopeType(
+          emptyScope.builder("Foo").set("x", IntType).set("y", IntType).set("z", BooleanType)
+            .build()
+        )
+      ),
+      scope
+    )
+  }
+
+  @Test
+  fun check_let_expression() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val scope = emptyScope.builder().apply {
+      set("numbers", SeqType(IntType))
+    }.build()
+
+    checkTypeIs(
+      PaxelParser.parse("from p in numbers let x = (p > 5) select x"),
+      SeqType(BooleanType),
+      scope
+    )
+  }
+
+  @Test
+  fun check_orderby_expression() {
+    val emptyScope = MapScope<InferredType>("root", mapOf())
+    val scope = emptyScope.builder().apply {
+      set("numbers", SeqType(IntType))
+    }.build()
+
+    checkTypeIs(
+      PaxelParser.parse("from p in numbers orderby p select p"),
+      SeqType(IntType),
+      scope
+    )
+
+    checkTypeIs(
+      PaxelParser.parse("from p in numbers orderby numbers select p"),
+      SeqType(IntType),
+      scope,
+      errors = listOf("order by expression numbers must be a primitive type" +
+                        " but was SeqType(PrimitiveType(Int)).")
```

Comment:
Nit: I would consider rendering these types in what I consider a friendlier way, e.g. instead of `SeqType(PrimitiveType(Int))` I would maybe print `Sequence<Int>`, up to you really.

For extra extra points, we could even print `UnionType(X, PrimitiveType(NullType))` as `X?`, though that's a bit more work, still I would argue it would be worthwhile.

Reply:
Did part 1, extra points save for later, because there are tricky cases:

`A | B?`  (e.g. `Int | String?`) is `UnionType(IntType, UnionType(StringType, NullType))`

vs

`A | B | null` `UnionType(IntType, StringType, NullType)` which arguably should happen but I don't want to add extra complex logic in an already sizable PR.